introspector/src/pc/mod.rs

```diff
@@ -216,57 +217,69 @@ impl ParachainCommander {
 
 	async fn watch_node_broadcast(
 		self,
-		from_collector: Receiver<CollectorUpdateEvent>,
+		mut from_collector: Receiver<CollectorUpdateEvent>,
 		api_service: CollectorStorageApi,
 	) {
-		let mut trackers: HashMap<u32, SubxtTracker> = HashMap::new();
+		let mut trackers: HashMap<u32, Arc<Mutex<SubxtTracker>>> = HashMap::new();
```

Comment:
Not here, it is the perfect use case for `Arc<Mutex<T>>` where it does not create any contention but provide safety guarantees.

Reply:
I am not a fan of Arc<Mutex<T>> with async code. Acquiring an unclaimed mutex I think is not really for free, but the cost is negligible. Can we implement something without Mutex ?

Reply:
We can actually remove the arc mutex if we one task per tracker. We would instead have a channel to send work to that task. Then repalce `Arc<Mutex<SubxtTracker>>` with Sender<TrackerUpdateMessage> in the hashmap. `TrackerUpdateMessage` is `(hash, block_number)`. The task should exit when channel closes and parent watcher task should remove the channel from the hashmap once that para is considered evicted.

Reply:
Furthermore, I actually think that the architecture with additional messages passing proxy will be less efficient and slower (due to 2 extra copying of each message) comparing to the case when each tracker lives in it's own `Arc<Mutex...>`. The only real case, where we have some contention is during a session update. I agree, that in that place we lock all trackers sequentially and it thus breaks parallel execution. However, this event happens very rarely and it's a negligible price for a code simplicity and efficiency when processing parachain updates.

Reply:
I think it is generally a pattern we want to avoid in async code. We also have an issue to remove such occurrences https://github.com/paritytech/polkadot-introspector/issues/73

Reply:
No, the pattern we want to avoid is a contention causing pattern, such as `Arc<Mutex<HashMap...>>`, where a shared state is accessed from the multiple futures. In this case, each tracker has no shared state with other trackers, so there is no contention. The reason why it is needed is to allow a shared storage of all trackers, but this storage itself is not under `Mutex` at all. 

So the pattern of the events processing is the following: we receive a message from the collector for a specific parachain, we spawn a future that works with this parachain and pass this message to this future. Afterwards, we go to the next parachain that lives in another not related `Arc<Mutex>>` cell and so on. Hence, `Arc<Mutex>>` pattern has no negative effect on asynchronous processing in this particular case.